rfcs/0019-storing-client-data.md

```diff
@@ -0,0 +1,648 @@
+
+For adding arbitrary transactions, the client could simply add it to the `transactions` database with `is_manually_added=true`. We can add arbitrary `receipt_chain_hashes` along with its parents to the `receipt_chain` database. The user can add arbitrary blocks along with their transactions into the `blocks_to_transactions` table, `block` and `transactions` database. When we add a block whose block length is greater than root length, the block will be added through the `transition_frontier_controller`. Namely, the block will be added to the `network_transition_writer` pipe in `Coda_networking.ml`.
+
+<a href="bootstrap"></a>
+
+### Running Bootstrap
+
+If a client is offline for long time and they have to bootstrap, then they would not have a `transition_frontier`. Therefore, all the transitions in the `transition_frontier` would have their `consensus_status` be considered to `UNKOWN`.
+
+## Consistency Issues
+
+As mentioned in the [Running Bootstrap section](#bootstrap), we would not be able to fully determine the `consensus_status` of the block that were in the transition_frontier. Worse, we would not be able to determine which blocks became snarked.
+
+We can have a third-party archived node tell a client the consensus_state of the blocks. Another option that we have is that we can create a P2P RPC call where peers can give a client a Merkle list proof of a state_hash and all the state_hashes of the blocks on top of the state_hash. The Merkle list should be length `K`, the number of block confirmations to gaurantee finality with a very high probability. If the Merkle list proof is size `K`, then the block should be fully confirmed. Otherwise, the block failed.
+
+# Rationale and Alternatives
+
+- Makes it easy to perform difficult queries for our protocol
+- Makes the client code more module
+- The current implementation takes a lot of work and it is quite limited for a certain use case.
+- Other alternatives are discussed in the previous commits of this [RFC](https://github.com/CodaProtocol/coda/pull/2901)
+- In an ideal world where we have an infinite amount of time, we would implement our own native graph database leverage the `transition_frontier` as an in-memory cache. Namely, we would have all the blocks in the `transition_frontier` to be in the in-memory cache along with nodes that have a 1 to 2 degree of seperation
+
+# Prior Art
+
+Ethereum is representing their data as patricia tries and are storing them in LevelDB.
+
+Coinbase also uses MongoDB to store all of the data in a blockchain.
+
+# Unresolved questions
+
+- How can we extend this design so that a client has multiple devices (i.e. Desktop application, mobile, etc...).
+
+# Appendix
+
+The flexibility of this new design changed some parts of the schema of the GraphqQL database. Here is the schema:
+
+```graphql
+enum ConsensusStatus {
+  FAILED
+  CONFIRMED
+  PENDING of int
```

Comment:
I don't think this (enums with payloads) is a thing in graphql. Also it's a little hard to scan this new schema to identify the changes.

Reply:
Yes enums with payloads is not a thing in graphql, can you adjust this?

Also can you rewrite this to only show the changes to the existing schema?